src/theory/quantifiers/ho_trigger.h

```diff
@@ -0,0 +1,121 @@
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief higher-order trigger class
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef __CVC4__THEORY__QUANTIFIERS__HO_TRIGGER_H
+#define __CVC4__THEORY__QUANTIFIERS__HO_TRIGGER_H
+
+#include <map>
+
+#include "expr/node.h"
+#include "theory/quantifiers/inst_match.h"
+#include "options/quantifiers_options.h"
+#include "theory/quantifiers/trigger.h"
+
+namespace CVC4 {
+namespace theory {
+namespace inst {
+
+class Trigger;
+
+/** HigherOrder trigger 
+*
+* This extends the trigger class with techniques that post-process instantiations, specified by
+* InstMatch objects, according to higher-order preunification (also called Huet's algorithm). 
+*
+* The main difference between HigherOrderTrigger and Trigger is the function sendInstantiation(...).
+* Recall that this function is called when its underlying IMGenerator generates
+* an InstMatch m using E-matching technique. We enumerate additional instantiations based on m
+* when the domain of m contains function variables.
+*
+* For examples, see description in the cpp.
+*/
+class HigherOrderTrigger : public Trigger {
+  friend class Trigger;
+private:
+  HigherOrderTrigger( QuantifiersEngine* qe, Node q, std::vector< Node >& nodes, 
+                      std::map< Node, std::vector< Node > >& ho_apps );
+  virtual ~HigherOrderTrigger();
+public:
+  /** collect all top-level HO_APPLY terms in n whose head is a variable in quantified formula q, store in apps */
+  static void collectHoVarApplyTerms( Node q, TNode n, std::map< Node, std::vector< Node > >& apps );
+  /** collect all top-level HO_APPLY terms in terms ns whose head is a variable in quantified formula q, store in apps */
+  static void collectHoVarApplyTerms( Node q, std::vector< Node >& ns, std::map< Node, std::vector< Node > >& apps );  
+  /** add all available instantiations exhaustively, extends Trigger::addInstantiations for this trigger */
+  virtual int addInstantiations( InstMatch& baseMatch );
+protected: 
+  /** map from variable functions to their applications in the quantifier d_f (member of Trigger) */
+  std::map< Node, std::vector< Node > > d_ho_var_apps;
+  /** list of all variable functions that are applied in d_f */
+  std::vector< Node > d_ho_var_list;
+  /** bound variables, bound variable list for each ho variable, used for constructing lambda terms in instantiations */
+  std::map< TNode, std::vector< Node > > d_ho_var_bvs;
+  std::map< TNode, Node > d_ho_var_bvl;
+  /** the set of types of ho variables */
+  std::vector< TypeNode > d_ho_var_types;
+  /** add higher-order type predicate lemmas 
+  * This predicate is used to force certain functions f of type tn to appear as first-class representatives in the
+  * quantifier-free UF solver. For a typical use case, we call getHoTypeMatchPredicate which returns a fresh 
```

Comment:
Where does tn come from?  What is the criteria for choosing "certain functions f"?